File: CSharp/SpaceVIL/SpaceVIL/Items/TitleBar.cs
using System;
using System.Drawing;
using SpaceVIL.Common;
using SpaceVIL.Core;
using SpaceVIL.Decorations;

namespace SpaceVIL
{
    /// <summary>
    /// TitleBar is the basic implementation of a user interface tittle bar of window.
    /// <para/> Extended from SpaceVIL.WindowAnchor. WindowAnchor is class representing the draggable window type of an item.
    /// <para/> Contains icon, text, minimize button, maximize button, close button.
    /// <para/> Supports all events except drag and drop despite that this class is draggable type.
    /// </summary>
    public class TitleBar : WindowAnchor
    {
        static int count = 0;
        private HorizontalStack _layout;

        /// <summary>
        /// Direction of title bar (in Mac OS is HorizontalDirection.FromRightToLeft, 
        /// in others OS HorizontalDirection.FromLeftToRight).
        /// </summary>
        public HorizontalDirection Direction = HorizontalDirection.FromLeftToRight;
        private Label _textLabel;
        private ImageItem _icon;

        /// <summary>
        /// Getting icon of title bar.
        /// </summary>
        /// <returns>Icon of title bar as SpaceVIL.ImageItem.</returns>
        public ImageItem GetIcon()
        {
            return _icon;
        }

        private ButtonCore _close;

        /// <summary>
        /// Getting close button of title bar. This button closes the current window.
        /// </summary>
        /// <returns>Close button of title bar as SpaceVIL.TitleBar.</returns>
        public ButtonCore GetCloseButton()
        {
            return _close;
        }

        private ButtonCore _minimize;

        /// <summary>
        /// Getting minimize button of title bar. This button minimizes the window to the taskbar.
        /// </summary>
        /// <returns>Minimize button of title bar as SpaceVIL.TitleBar.</returns>
        public ButtonCore GetMinimizeButton()
        {
            return _minimize;
        }

        private ButtonCore _maximize;

        /// <summary>
        /// Getting maximize button of title bar. This button maximizes/restores the 
        /// window to all available space of the current display 
        /// (display size without task bar and other OS elements).
        /// </summary>
        /// <returns>Maximize button of title bar as SpaceVIL.TitleBar.</returns>
        public ButtonCore GetMaximizeButton()
        {
            return _maximize;
        }

        /// <summary>
        /// Default TitleBar constructor.
        /// </summary>
        public TitleBar()
        {
            SetItemName("TitleBar_" + count);
            count++;

            _layout = new HorizontalStack();
            _textLabel = new Label();
            _textLabel.IsFocusable = false;
            _minimize = new ButtonCore();
            _minimize.IsFocusable = false;
            _maximize = new ButtonCore();
            _maximize.IsFocusable = false;
            _close = new ButtonCore();
            _close.IsFocusable = false;
            _icon = new ImageItem();
            _icon.IsFocusable = false;

            SetStyle(DefaultsService.GetDefaultStyle(typeof(SpaceVIL.TitleBar)));

            EventMouseDoubleClick += (sender, args) =>
            {
                GetHandler().Maximize();
            };
        }

        /// <summary>
        /// Constructs TitleBar with the specified title text.
        /// </summary>
        /// <param name="text">Title text</param>
        public TitleBar(String text) : this()
        {
            SetText(text);
        }

        /// <summary>
        /// Setting image icon for title bar. The image is scaled to the specified width and height.
        /// </summary>
        /// <param name="icon">Image icon as System.Drawing.Bitmap.</param>
        /// <param name="width">New width of the bitmap.</param>
        /// <param name="height">New height of the bitmap.</param>
        public void SetIcon(Bitmap icon, int width, int height)
        {
            _icon.SetSize(width, height);
            _icon.SetImage(icon);
            _icon.SetVisible(true);
        }

        /// <summary>
        /// Setting alignment of a TitleBar text. 
        /// Combines with alignment by vertically (Top, VCenter, Bottom) and horizontally (Left, HCenter, Right). 
        /// </summary>
        /// <param name="alignment">Text alignment as SpaceVIL.Core.ItemAlignment.</param>
        public void SetTextAlignment(ItemAlignment alignment)
        {
            _textLabel.SetTextAlignment(alignment);
        }

        /// <summary>
        /// Setting alignment of a TitleBar text. 
        /// Combines with alignment by vertically (Top, VCenter, Bottom) and horizontally (Left, HCenter, Right). 
        /// </summary>
        /// <param name="alignment">Text alignment as sequence of SpaceVIL.Core.ItemAlignment.</param>
        public void SetTextAlignment(params ItemAlignment[] alignment)
        {
            _textLabel.SetTextAlignment(alignment);
        }

        /// <summary>
        /// Setting indents for the text to offset text relative to TitleBar.
        /// </summary>
        /// <param name="margin">Indents as SpaceVIL.Decorations.Indents.</param>
        public void SetTextMargin(Indents margin)
        {
            _textLabel.SetMargin(margin);
        }

        /// <summary>
        /// Setting indents for the text to offset text relative to TitleBar.
        /// </summary>
        /// <param name="left">Indent on the left.</param>
        /// <param name="top">Indent on the top.</param>
        /// <param name="right">Indent on the right.</param>
        /// <param name="bottom">Indent on the bottom.</param>
        public void SetTextMargin(int left = 0, int top = 0, int right = 0, int bottom = 0)
        {
            _textLabel.SetMargin(left, top, right, bottom);
        }

        /// <summary>
        /// Getting indents of the text.
        /// </summary>
        /// <returns>Indents as SpaceVIL.Decorations.Indents.</returns>
        public Indents GetTextMargin()
        {
            return _textLabel.GetMargin();
        }

        /// <summary>
        /// Setting font of the text.
        /// </summary>
        /// <param name="font">Font as System.Drawing.Font.</param>
        public void SetFont(Font font)
        {
            _textLabel.SetFont(font);
        }

        /// <summary>
        /// Setting font size of the text.
        /// </summary>
        /// <param name="size">New size of the font.</param>
        public void SetFontSize(int size)
        {
            _textLabel.SetFontSize(size);
        }

        /// <summary>
        /// Setting font style of the text.
        /// </summary>
        /// <param name="style">New font style as System.Drawing.FontStyle.</param>
        public void SetFontStyle(FontStyle style)
        {
            _textLabel.SetFontStyle(style);
        }

        /// <summary>
        /// Setting new font family of the text.
        /// </summary>
        /// <param name="fontFamily">New font family as System.Drawing.FontFamily.</param>
        public void SetFontFamily(FontFamily fontFamily)
        {
            _textLabel.SetFontFamily(fontFamily);
        }

        /// <summary>
        /// Getting the current font of the text.
        /// </summary>
        /// <returns>Font as System.Drawing.Font.</returns>
        public Font GetFont()
        {
            return _textLabel.GetFont();
        }

        /// <summary>
        /// Setting the text.
        /// </summary>
        /// <param name="text">Text as System.String.</param>
        public virtual void SetText(String text)
        {
            _textLabel.SetText(text);
        }

        /// <summary>
        /// Getting the current text of the TitleBar.
        /// </summary>
        /// <returns>Text as System.String.</returns>
        public virtual String GetText()
        {
            return _textLabel.GetText();
        }

        /// <summary>
        /// Getting the text width (useful when you need resize TitleBar by text width).
        /// </summary>
        /// <returns>Text width.</returns>
        public int GetTextWidth()
        {
            return _textLabel.GetWidth();
        }

        /// <summary>
        /// Getting the text height (useful when you need resize TitleBar by text height).
        /// </summary>
        /// <returns>Text height.</returns>
        public int GetTextHeight()
        {
            return _textLabel.GetHeight();
        }

        /// <summary>
        /// Setting text color of a TitleBar.
        /// </summary>
        /// <param name="color">Text color as System.Drawing.Color.</param>
        public void SetForeground(Color color)
        {
            _textLabel.SetForeground(color);
        }

        /// <summary>
        /// Setting text color of a TitleBar in byte RGB format.
        /// </summary>
        /// <param name="r">Red color component. Range: (0 - 255)</param>
        /// <param name="g">Green color component. Range: (0 - 255)</param>
        /// <param name="b">Blue color component. Range: (0 - 255)</param>
        public void SetForeground(int r, int g, int b)
        {
            _textLabel.SetForeground(r, g, b);
        }

        /// <summary>
        /// Setting text color of a TitleBar in byte RGBA format.
        /// </summary>
        /// <param name="r">Red color component. Range: (0 - 255)</param>
        /// <param name="g">Green color component. Range: (0 - 255)</param>
        /// <param name="b">Blue color component. Range: (0 - 255)</param>
        /// <param name="a">Alpha color component. Range: (0 - 255)</param>
        public void SetForeground(int r, int g, int b, int a)
        {
            _textLabel.SetForeground(r, g, b, a);
        }

        /// <summary>
        /// Setting text color of a TitleBar in float RGB format.
        /// </summary>
        /// <param name="r">Red color component. Range: (0.0f - 1.0f)</param>
        /// <param name="g">Green color component. Range: (0.0f - 1.0f)</param>
        /// <param name="b">Blue color component. Range: (0.0f - 1.0f)</param>
        public void SetForeground(float r, float g, float b)
        {
            _textLabel.SetForeground(r, g, b);
        }

        /// <summary>
        /// Setting text color of a TitleBar in float RGBA format.
        /// </summary>
        /// <param name="r">Red color component. Range: (0.0f - 1.0f)</param>
        /// <param name="g">Green color component. Range: (0.0f - 1.0f)</param>
        /// <param name="b">Blue color component. Range: (0.0f - 1.0f)</param>
        /// <param name="a">Alpha color component. Range: (0.0f - 1.0f)</param>
        public void SetForeground(float r, float g, float b, float a)
        {
            _textLabel.SetForeground(r, g, b, a);
        }

        /// <summary>
        /// Getting current text color.
        /// </summary>
        /// <returns>Text color as System.Drawing.Color.</returns>
        public Color GetForeground()
        {
            return _textLabel.GetForeground();
        }

        /// <summary>
        /// Initializing all elements in the TitleBar.
        /// <para/> Notice: This method is mainly for overriding only. SpaceVIL calls 
        /// this method if necessary and no need to call it manually.
        /// </summary>
        public override void InitElements()
        {
            AddItem(_layout);

            //_close
            _close.EventMouseClick += (sender, args) =>
            {
                GetHandler().EventClose?.Invoke();
            };

            //_minimize
            _minimize.EventMouseClick += (sender, args) =>
            {
                GetHandler().Minimize();
            };

            //_maximize
            _maximize.EventMouseClick += (sender, args) =>
            {
                GetHandler().Maximize();
            };

            //adding
            switch (Direction)
            {
                case HorizontalDirection.FromLeftToRight:
                    _layout.AddItems(_icon, _textLabel, _minimize, _maximize, _close);
                    break;
                case HorizontalDirection.FromRightToLeft:
                    _textLabel.SetTextAlignment(ItemAlignment.HCenter | ItemAlignment.VCenter);
                    _layout.AddItems(_close, _minimize, _maximize, _icon, _textLabel);
                    break;
                default:
                    _layout.AddItems(_icon, _textLabel, _minimize, _maximize, _close);
                    break;
            }
        }

        /// <summary>
        /// Setting style of the TitleBar. 
        /// <para/> Inner styles: "closebutton", "minimizebutton", "maximizebutton", "title".
        /// </summary>
        /// <param name="style">Style as SpaceVIL.Decorations.Style.</param>
        public override void SetStyle(Style style)
        {
            if (style == null)
                return;

            base.SetStyle(style);
            SetForeground(style.Foreground);
            SetFont(style.Font);
            SetTextAlignment(style.TextAlignment);
            _layout.SetSpacing(style.Spacing);

            Style innerStyle = style.GetInnerStyle("closebutton");
            if (innerStyle != null)
            {
                _close.SetStyle(innerStyle);
            }
            innerStyle = style.GetInnerStyle("minimizebutton");
            if (innerStyle != null)
            {
                _minimize.SetStyle(innerStyle);
            }
            innerStyle = style.GetInnerStyle("maximizebutton");
            if (innerStyle != null)
            {
                _maximize.SetStyle(innerStyle);
            }
            innerStyle = style.GetInnerStyle("title");
            if (innerStyle != null)
            {
                SetTextMargin(innerStyle.Margin);
            }

            //icon
            _icon.SetVisible(false);
            _icon.SetBackground(Color.Transparent);
            _icon.SetSizePolicy(SizePolicy.Fixed, SizePolicy.Fixed);
            _icon.SetAlignment(ItemAlignment.VCenter | ItemAlignment.Left);
        }
    }
}

File: Java/SpaceVIL/spacevil/src/main/java/com/spvessel/spacevil/TitleBar.java
package com.spvessel.spacevil;

import com.spvessel.spacevil.Flags.HorizontalDirection;
import com.spvessel.spacevil.Common.DefaultsService;
import com.spvessel.spacevil.Decorations.Indents;
import com.spvessel.spacevil.Decorations.Style;
import com.spvessel.spacevil.Flags.ItemAlignment;
import com.spvessel.spacevil.Flags.SizePolicy;

import java.awt.*;
import java.awt.image.BufferedImage;
import java.util.List;

/**
 * TitleBar is the basic implementation of a user interface tittle bar of
 * window.
 * <p>
 * Extended from com.spvessel.spacevil.WindowAnchor. WindowAnchor is class
 * representing the draggable window type of an item.
 * <p>
 * Contains icon, text, minimize button, maximize button, close button.
 * <p>
 * Supports all events except drag and drop despite that this class is draggable
 * type.
 */
public class TitleBar extends WindowAnchor {

    private static int count = 0;
    private HorizontalStack _layout;

    /**
     * Direction of title bar (in Mac OS is HorizontalDirection.FROM_LEFT_TO_RIGHT,
     * in others OS HorizontalDirection.FROM_RIGHT_TO_LEFT).
     */
    public HorizontalDirection direction = HorizontalDirection.FROM_LEFT_TO_RIGHT;

    private Label _texLabel;
    private ImageItem _icon;

    /**
     * Getting icon of title bar.
     * 
     * @return Icon of title bar as com.spvessel.spacevil.ImageItem.
     */
    public ImageItem getIcon() {
        return _icon;
    }

    private ButtonCore _close;

    /**
     * Getting close button of title bar. This button closes the current window.
     * 
     * @return Close button of title bar as com.spvessel.spacevil.TitleBar.
     */
    public ButtonCore getCloseButton() {
        return _close;
    }

    private ButtonCore _minimize;

    /**
     * Getting minimize button of title bar. This button minimizes the window to the
     * taskbar.
     * 
     * @return Minimize button of title bar as com.spvessel.spacevil.TitleBar.
     */
    public ButtonCore getMinimizeButton() {
        return _minimize;
    }

    private ButtonCore _maximize;

    /**
     * Getting maximize button of title bar. This button maximizes/restores the
     * window to all available space of the current display (display size without
     * task bar and other OS elements).
     * 
     * @return Maximize button of title bar as com.spvessel.spacevil.TitleBar.
     */
    public ButtonCore getMaximizeButton() {
        return _maximize;
    }

    /**
     * Default TitleBar constructor.
     */
    public TitleBar() {
        setItemName("TitleBar_" + count);
        count++;

        _layout = new HorizontalStack();
        _texLabel = new Label();
        _texLabel.isFocusable = false;
        _minimize = new ButtonCore();
        _minimize.isFocusable = false;
        _maximize = new ButtonCore();
        _maximize.isFocusable = false;
        _maximize.setItemName("MAXIMIZE");
        _close = new ButtonCore();
        _close.isFocusable = false;
        _icon = new ImageItem();
        _icon.isFocusable = false;

        setStyle(DefaultsService.getDefaultStyle(TitleBar.class));

        eventMouseDoubleClick.add((sender, args) -> {
            getHandler().maximize();
        });
    }

    /**
     * Constructs TitleBar with the specified title text.
     * 
     * @param text Title text
     */
    public TitleBar(String text) {
        this();
        setText(text);
    }

    /**
     * Setting image icon for title bar. The image is scaled to the specified width
     * and height.
     * 
     * @param icon   Image icon as java.awt.image.BufferedImage.
     * @param width  New width of the bitmap.
     * @param height New height of the bitmap.
     */
    public void setIcon(BufferedImage icon, int width, int height) {
        _icon.setSize(width, height);
        _icon.setImage(icon);
        _icon.setVisible(true);
    }

    /**
     * Setting alignment of a TitleBar text. Combines with alignment by vertically
     * (TOP, VCENTER, BOTTOM) and horizontally (LEFT, HCENTER, RIGHT).
     * 
     * @param alignment Text alignment as com.spvessel.spacevil.Flags.ItemAlignment.
     */
    public void setTextAlignment(ItemAlignment... alignment) {
        _texLabel.setTextAlignment(alignment);
    }

    /**
     * Setting alignment of a TitleBar text. Combines with alignment by vertically
     * (TOP, VCENTER, BOTTOM) and horizontally (LEFT, HCENTER, RIGHT).
     * 
     * @param alignment Text alignment as List of
     *                  com.spvessel.spacevil.Flags.ItemAlignment.
     */
    public void setTextAlignment(List<ItemAlignment> alignment) {
        _texLabel.setTextAlignment(alignment);
    }

    /**
     * Setting indents for the text to offset text relative to TitleBar.
     * 
     * @param margin Indents as com.spvessel.spacevil.Decorations.Indents.
     */
    public void setTextMargin(Indents margin) {
        _texLabel.setMargin(margin);
    }

    /**
     * Setting indents for the text to offset text relative to TitleBar.
     * 
     * @param left   Indent on the left.
     * @param top    Indent on the top.
     * @param right  Indent on the right.
     * @param bottom Indent on the bottom.
     */
    public void setTextMargin(int left, int top, int right, int bottom) {
        _texLabel.setMargin(left, top, right, bottom);
    }

    /**
     * Getting indents of the text.
     * 
     * @return Indents as com.spvessel.spacevil.Decorations.Indents.
     */
    public Indents getTextMargin() {
        return _texLabel.getMargin();
    }

    /**
     * Setting font of the text.
     * 
     * @param font Font as java.awt.Font.
     */
    public void setFont(Font font) {
        _texLabel.setFont(font);
    }

    /**
     * Setting font size of the text.
     * 
     * @param size New size of the font.
     */
    public void setFontSize(int size) {
        _texLabel.setFontSize(size);
    }

    /**
     * Setting font style of the text.
     * 
     * @param style New font style (from java.awt.Font package).
     */
    public void setFontStyle(int style) {
        _texLabel.setFontStyle(style);
    }

    /**
     * Setting new font family of the text.
     * 
     * @param fontFamily New font family name.
     */
    public void setFontFamily(String fontFamily) {
        _texLabel.setFontFamily(fontFamily);
    }

    /**
     * Getting the current font of the text.
     * 
     * @return Font as java.awt.Font.
     */
    public Font getFont() {
        return _texLabel.getFont();
    }

    /**
     * Setting the text.
     * 
     * @param text Text as java.lang.String.
     */
    public void setText(String text) {
        _texLabel.setText(text);
    }

    /**
     * Getting the current text of the TitleBar.
     * 
     * @return Text as java.lang.String.
     */
    public String getText() {
        return _texLabel.getText();
    }

    /**
     * Getting the text width (useful when you need resize TitleBar by text width).
     * 
     * @return Text width.
     */
    public int getTextWidth() {
        return _texLabel.getWidth();
    }

    /**
     * Getting the text height (useful when you need resize TitleBar by text
     * height).
     * 
     * @return Text height.
     */
    public int getTextHeight() {
        return _texLabel.getHeight();
    }

    /**
     * Setting text color of a TitleBar.
     * 
     * @param color Text color as java.awt.Color.
     */
    public void setForeground(Color color) {
        _texLabel.setForeground(color);
    }

    /**
     * Setting text color of a TitleBar in byte RGB format.
     * 
     * @param r Red color component. Range: (0 - 255)
     * @param g Green color component. Range: (0 - 255)
     * @param b Blue color component. Range: (0 - 255)
     */
    public void setForeground(int r, int g, int b) {
        _texLabel.setForeground(r, g, b);
    }

    /**
     * Setting background color of an item in byte RGBA format.
     * 
     * @param r Red color component. Range: (0 - 255)
     * @param g Green color component. Range: (0 - 255)
     * @param b Blue color component. Range: (0 - 255)
     * @param a Alpha color component. Range: (0 - 255)
     */
    public void setForeground(int r, int g, int b, int a) {
        _texLabel.setForeground(r, g, b, a);
    }

    /**
     * Setting text color of a TitleBar in float RGB format.
     * 
     * @param r Red color component. Range: (0.0f - 1.0f)
     * @param g Green color component. Range: (0.0f - 1.0f)
     * @param b Blue color component. Range: (0.0f - 1.0f)
     */
    public void setForeground(float r, float g, float b) {
        _texLabel.setForeground(r, g, b);
    }

    /**
     * Setting text color of a TitleBar in float RGBA format.
     * 
     * @param r Red color component. Range: (0.0f - 1.0f)
     * @param g Green color component. Range: (0.0f - 1.0f)
     * @param b Blue color component. Range: (0.0f - 1.0f)
     * @param a Alpha color component. Range: (0.0f - 1.0f)
     */
    public void setForeground(float r, float g, float b, float a) {
        _texLabel.setForeground(r, g, b, a);
    }

    /**
     * Getting current text color.
     * 
     * @return Text color as as java.awt.Color.
     */
    public Color getForeground() {
        return _texLabel.getForeground();
    }

    /**
     * Initializing all elements in the TitleBar.
     * <p>
     * Notice: This method is mainly for overriding only. SpaceVIL calls this method
     * if necessary and no need to call it manually.
     */
    @Override
    public void initElements() {
        addItem(_layout);

        // _close
        _close.eventMouseClick.add((sender, args) -> {
            getHandler().eventClose.execute();
        });

        // _minimize
        _minimize.eventMouseClick.add((sender, args) -> {
            getHandler().minimize();
        });

        // _maximize
        _maximize.eventMouseClick.add((sender, args) -> {
            getHandler().maximize();
        });

        // adding
        switch (direction) {
            case FROM_LEFT_TO_RIGHT:
                _layout.addItems(_icon, _texLabel, _minimize, _maximize, _close);
                break;
            case FROM_RIGHT_TO_LEFT:
                _texLabel.setTextAlignment(ItemAlignment.HCENTER, ItemAlignment.VCENTER);
                _layout.addItems(_close, _minimize, _maximize, _icon, _texLabel);
                break;
            default:
                _layout.addItems(_icon, _texLabel, _minimize, _maximize, _close);
                break;
        }
        _minimize.setPassEvents(false);
        _maximize.setPassEvents(false);
        _close.setPassEvents(false);
    }

    /**
     * Setting style of the TitleBar.
     * <p>
     * Inner styles: "closebutton", "minimizebutton", "maximizebutton", "title".
     * 
     * @param style Style as com.spvessel.spacevil.Decorations.Style.
     */
    @Override
    public void setStyle(Style style) {

        if (style == null)
            return;

        super.setStyle(style);
        setForeground(style.foreground);
        setFont(style.font);
        setTextAlignment(style.textAlignment);
        _layout.setSpacing(style.spacing);

        Style inner_style = style.getInnerStyle("closebutton");
        if (inner_style != null) {
            _close.setStyle(inner_style);
        }
        inner_style = style.getInnerStyle("minimizebutton");
        if (inner_style != null) {
            _minimize.setStyle(inner_style);
        }
        inner_style = style.getInnerStyle("maximizebutton");
        if (inner_style != null) {
            _maximize.setStyle(inner_style);
        }
        inner_style = style.getInnerStyle("title");
        if (inner_style != null) {
            setTextMargin(inner_style.margin);
        }

        _icon.setVisible(false);
        _icon.setBackground(0, 0, 0, 0);
        _icon.setSizePolicy(SizePolicy.FIXED, SizePolicy.FIXED);
        _icon.setAlignment(ItemAlignment.VCENTER, ItemAlignment.LEFT);
    }
}
